proposals/xxxx-package-manager-performance-testing.md

```diff
@@ -0,0 +1,63 @@
+# Package Manager Performance Testing Conventions
+
+* Proposal: [SE-XXXX](xxx-package-manager-performance-testing.md)
+* Author: [Ankit Aggarwal](https://github.com/aciidb0mb3r)
+* Review Manager: TBD
+* Status: Discussion
+
+## Introduction
+
+This is a proposal for adding package manager support for building and running performance tests in release mode.
+
+## Motivation
+
+The package manager supports building and running tests using the `swift test` subtool. The sources are built in debug mode with swift compiler's
+`-enable-testing` flag allowing testable imports in test modules (which gives it internal level access to the imported module). 
+This works well for normal unit tests, however for performance tests we are usually interested in release build test results which takes advantages of compiler 
+optimizations.
+
+We can provide an option to compile and run the tests in release mode with `-enable-testing` flag. However that will disable some of the compiler optimizations, which 
+may be of interest to packages having very performance sensitive code (which critically depends on those optimization).
+
+Since performance tests usually take more time to run than other tests, there is also a requirement to have a way to run performance tests separate from unit tests.
+As a concrete example, package manager's own performance tests are disabled from running on Swift CI.
+
+## Proposed solution
+
+We propose to extend the test modules conventions and consider a directory as "performance test module" which has a suffix "PerformanceTests". The tests in these modules will only run when `--performance` flag is passed to `swift test`.
+
+This command will run performance tests in debug configuration:
+
+```
+$ swift test --performance
+```
+
+And this command will run performance tests in release configuration:
+
+```
+$ swift test --performance -c release
+```
+
+It may be desirable to run performance test in release mode with testing enabled. We propose to add a new boolean property (`enablePerformanceTestability`) in manifest to turn on testability for performance tests in release mode, so packages which want to use the feature in performance tests can do so without passing `-Xswiftc -enable-testing` flags. The default value of this property will be `false`:
+
+```swift
+let package = Package(
+    name: "MyPackage",
+    enablePerformanceTestability: true 
+)
+```
+
+## Detailed Design
+
+Currently one XCTest product (bundle on macOS, executable on Linux) is created and run for all the test modules with name `<package-name>PackageTests.xctest`.
+
+We propose to create another test product with name `<package-name>PackagePerformanceTests.xctest` which will contain only performance tests.
+This will allow package manager to run these test products separately.
+
+## Impact on existing code
+
+There will be no impact on exisiting code as this is purely additive.
+
+## Alternative considered
+
+None at this time.
```

Comment:
We should fill this in with other alternatives, like:

1. Do nothing special, just suggest users use XCTest features as normal. SwiftPM will have no ability to slice and dice out performance tests.

2. Build an XCTest mechanism which solves the con of #1 by allowing us to simply disable all performance test blocks.

3. Add a manifest flag for "requiresTestability", but nothing else. Allows users to build their own conventions (using, e.g., Swift language configuration checks).

4. We could have defined a new configuration which means "release + testability" and then the `test` command just always runs the most tests available to it given the current configuration. This is probably the simplest model, but maybe gives up the ability for us to automatically do the build twice, a feature I like.